File: src/main/java/com/github/jnhyperion/hyperrobotframeworkplugin/ide/parameterinfo/ParameterInfoHandlerUtil.java
package com.github.jnhyperion.hyperrobotframeworkplugin.ide.parameterinfo;

import com.github.jnhyperion.hyperrobotframeworkplugin.psi.element.Argument;
import com.github.jnhyperion.hyperrobotframeworkplugin.psi.element.KeywordStatement;
import com.github.jnhyperion.hyperrobotframeworkplugin.psi.element.Parameter;
import com.intellij.lang.parameterInfo.ParameterInfoUIContext;
import com.intellij.lang.parameterInfo.ParameterInfoUIContextEx;
import com.intellij.openapi.util.Pair;
import com.intellij.psi.SyntaxTraverser;
import com.intellij.psi.tree.IElementType;
import com.jetbrains.python.psi.PyFunction;
import com.jetbrains.python.psi.PyNamedParameter;
import com.jetbrains.python.psi.PySingleStarParameter;
import com.jetbrains.python.psi.PySlashParameter;
import com.jetbrains.python.psi.PyTupleParameter;
import com.jetbrains.python.psi.impl.ParamHelper;
import com.jetbrains.python.psi.types.PyCallableParameter;
import com.jetbrains.python.psi.types.PyCallableParameterImpl;
import com.jetbrains.python.psi.types.PyCallableType;
import com.jetbrains.python.psi.types.PyStructuralType;
import com.jetbrains.python.psi.types.TypeEvalContext;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

public class ParameterInfoHandlerUtil {

    static <V> int getCurrentParameterIndex(SyntaxTraverser<V> traverser, int offset, IElementType... delimiterTypes) {
        V root = traverser.getRoot();
        int curOffset = traverser.api.rangeOf(root).getStartOffset();
        if (offset < curOffset) {
            return -1;
        }

        int index = 0;
        Set<IElementType> delimiterTypeSet = Set.of(delimiterTypes);
        for (V child : traverser) {
            curOffset += traverser.api.rangeOf(child).getLength();
            if (offset < curOffset) {
                break;
            }

            IElementType type = traverser.api.typeOf(child);
            if (delimiterTypeSet.contains(type)) {
                index++;
            }
        }

        return index;
    }

    static Pair<List<String>, List<String>> buildHintsAndAnnotationsWithHighlights(PyFunction callingFunction,
                                                                                   @NotNull ParameterInfoUIContext context,
                                                                                   int currentParamIndex,
                                                                                   Map<Integer, EnumSet<ParameterInfoUIContextEx.Flag>> hintFlags) {
        KeywordStatement keywordStatement = (KeywordStatement) context.getParameterOwner();
        Argument[] arguments = keywordStatement.getArguments().toArray(Argument[]::new);

        final Map<Integer, PyCallableParameter> indexToNamedParameter = new HashMap<>();

        // param -> hint index. indexes are not contiguous, because some hints are parentheses.
        final Map<PyCallableParameter, Integer> parameterToHintIndex = new HashMap<>();

        final TypeEvalContext typeEvalContext = TypeEvalContext.codeAnalysis(keywordStatement.getProject(), callingFunction.getContainingFile());
        final List<PyCallableParameter> parameters = callingFunction.getParameters(typeEvalContext);
        final PyCallableType callableType = (PyCallableType) typeEvalContext.getType(callingFunction);
        if (callableType == null) {
            return null;
        }

        final Pair<List<String>, List<String>> hintsAndAnnotations = ParameterInfoHandlerUtil.buildParameterListHint(parameters,
                                                                                                                     indexToNamedParameter,
                                                                                                                     parameterToHintIndex,
                                                                                                                     hintFlags,
                                                                                                                     typeEvalContext);
        ParameterInfoHandlerUtil.highlightParameters(Arrays.asList(arguments),
                                                     callableType,
                                                     parameters,
                                                     indexToNamedParameter,
                                                     parameterToHintIndex,
                                                     hintFlags,
                                                     currentParamIndex);

        return hintsAndAnnotations;
    }

    /**
     * Builds a list of textual representation of parameters
     * Returns two lists: parameters with type hints and parameters with type annotations
     *
     * @param parameters            parameters of a callable
     * @param indexToNamedParameter used to collect all named parameters of callable
     * @param parameterToHintIndex  used to collect info about parameter hints
     * @param hintFlags             mark parameter as deprecated/highlighted/strikeout
     * @param context               context to be used to get parameter representation
     */
    static Pair<List<String>, List<String>> buildParameterListHint(@NotNull List<PyCallableParameter> parameters,
                                                                   @NotNull final Map<Integer, PyCallableParameter> indexToNamedParameter,
                                                                   @NotNull final Map<PyCallableParameter, Integer> parameterToHintIndex,
                                                                   @NotNull final Map<Integer, EnumSet<ParameterInfoUIContextEx.Flag>> hintFlags,
                                                                   @NotNull TypeEvalContext context) {
        final List<String> hintsList = new ArrayList<>();
        final List<String> annotations = new ArrayList<>();
        final int[] currentParameterIndex = new int[] { 0 };
        ParamHelper.walkDownParameters(parameters, new ParamHelper.ParamWalker() {
            @Override
            public void enterTupleParameter(PyTupleParameter param, boolean first, boolean last) {
                hintFlags.put(hintsList.size(), EnumSet.noneOf(ParameterInfoUIContextEx.Flag.class));
                hintsList.add("(");
                annotations.add("");
            }

            @Override
            public void leaveTupleParameter(PyTupleParameter param, boolean first, boolean last) {
                hintFlags.put(hintsList.size(), EnumSet.noneOf(ParameterInfoUIContextEx.Flag.class));
                hintsList.add(last ? ")" : "), ");
                annotations.add("");
            }

            @Override
            public void visitNamedParameter(PyNamedParameter param, boolean first, boolean last) {
                visitNonPsiParameter(PyCallableParameterImpl.psi(param), first, last);
            }

            @Override
            public void visitSlashParameter(@NotNull PySlashParameter param, boolean first, boolean last) {
                hintFlags.put(hintsList.size(), EnumSet.noneOf(ParameterInfoUIContextEx.Flag.class));
                hintsList.add(last ? PySlashParameter.TEXT : (PySlashParameter.TEXT + ", "));
                annotations.add("");
                currentParameterIndex[0]++;
            }

            @Override
            public void visitSingleStarParameter(PySingleStarParameter param, boolean first, boolean last) {
                hintFlags.put(hintsList.size(), EnumSet.noneOf(ParameterInfoUIContextEx.Flag.class));
                hintsList.add(last ? PySingleStarParameter.TEXT : (PySingleStarParameter.TEXT + ", "));
                annotations.add("");
                currentParameterIndex[0]++;
            }

            @Override
            public void visitNonPsiParameter(@NotNull PyCallableParameter parameter, boolean first, boolean last) {
                indexToNamedParameter.put(currentParameterIndex[0], parameter);
                final StringBuilder stringBuilder = new StringBuilder();
                boolean annotationAdded = false;
                if (parameter.getParameter() instanceof PyNamedParameter) {
                    final String annotation = ((PyNamedParameter) parameter.getParameter()).getAnnotationValue();
                    if (annotation != null) {
                        String annotationText = ParamHelper.getNameInSignature(parameter) + ": " + annotation.replaceAll("\n", "").replaceAll("\\s+", " ");
                        annotations.add(last ? annotationText : (annotationText + ", "));
                        annotationAdded = true;
                    }
                }
                if (!annotationAdded) {
                    annotations.add("");
                }
                stringBuilder.append(parameter.getPresentableText(true, context, type -> type == null || type instanceof PyStructuralType));
                if (!last) {
                    stringBuilder.append(", ");
                }
                final int hintIndex = hintsList.size();
                parameterToHintIndex.put(parameter, hintIndex);
                hintFlags.put(hintIndex, EnumSet.noneOf(ParameterInfoUIContextEx.Flag.class));
                hintsList.add(stringBuilder.toString());
                currentParameterIndex[0]++;
            }
        });
        return new Pair<>(hintsList, annotations);
    }

    static void highlightParameters(List<Argument> arguments,
                                    PyCallableType callableType,
                                    List<PyCallableParameter> parameters,
                                    Map<Integer, PyCallableParameter> indexToNamedParameter,
                                    Map<PyCallableParameter, Integer> parameterToHintIndex,
                                    Map<Integer, EnumSet<ParameterInfoUIContextEx.Flag>> hintFlags,
                                    int currentParamIndex) {
        // gray out enough first parameters as implicit (self, cls, ...)
        for (int i = 0; i < callableType.getImplicitOffset(); i++) {
            if (indexToNamedParameter.containsKey(i)) {
                final PyCallableParameter parameter = indexToNamedParameter.get(i);
                hintFlags.get(parameterToHintIndex.get(parameter)).add(ParameterInfoUIContextEx.Flag.DISABLE); // show but mark as absent
            }
        }

        collectHighlights(callableType, parameters, parameterToHintIndex, hintFlags, arguments, currentParamIndex);
    }

    /**
     * match params to available args, highlight current param(s)
     */
    static void collectHighlights(@NotNull final PyCallableType callableType,
                                  @NotNull final List<PyCallableParameter> parameterList,
                                  @NotNull final Map<PyCallableParameter, Integer> parameterHintToIndex,
                                  @NotNull final Map<Integer, EnumSet<ParameterInfoUIContextEx.Flag>> hintFlags,
                                  @NotNull final List<Argument> args,
                                  int currentParamIndex) {
        Map<String, PyCallableParameter> nameToCallableParameterIndex = parameterList.stream()
                                                                                     .filter(param -> param.getName() != null)
                                                                                     .collect(Collectors.toMap(PyCallableParameter::getName, param -> param));

        int implicitOffset = callableType.getImplicitOffset();
        int positionalContainerIndex = findPositionalContainerIndex(parameterList);
        boolean parameterFound = false;
        for (int i = 0; i < args.size(); i++) {
            Argument arg = args.get(i);
            boolean mustHighlight = currentParamIndex == i;
            if (arg instanceof Parameter parameter) {
                parameterFound = true;

                String parameterName = parameter.getParameterName();
                if (nameToCallableParameterIndex.containsKey(parameterName)) {
                    PyCallableParameter pyCallableParameter = nameToCallableParameterIndex.get(parameterName);
                    highlightParameter(pyCallableParameter, parameterHintToIndex, hintFlags, mustHighlight);
                } else {
                    int lastParameterIndex = parameterList.size() - 1;
                    PyCallableParameter lastCallableParameter = parameterList.get(lastParameterIndex);
                    if (lastCallableParameter.isKeywordContainer()) {
                        highlightParameter(lastCallableParameter, parameterHintToIndex, hintFlags, mustHighlight);
                    }
                }
            } else {
                if (!parameterFound) {
                    int argIndex = Math.min(i + implicitOffset, positionalContainerIndex);
                    PyCallableParameter pyCallableParameter = parameterList.get(argIndex);
                    highlightParameter(pyCallableParameter, parameterHintToIndex, hintFlags, mustHighlight);
                }
            }
        }
    }

    private static int findPositionalContainerIndex(@NotNull List<PyCallableParameter> parameterList) {
        for (int i = 0; i < parameterList.size(); i++) {
            PyCallableParameter parameter = parameterList.get(i);
            if (parameter.isPositionalContainer()) {
                return i;
            }
        }
        return -1;
    }

    private static void highlightParameter(@NotNull final PyCallableParameter parameter,
                                           @NotNull final Map<PyCallableParameter, Integer> parameterToHintIndex,
                                           @NotNull final Map<Integer, EnumSet<ParameterInfoUIContextEx.Flag>> hintFlags,
                                           final boolean mustHighlight) {
        final Integer hintIndex = parameterToHintIndex.get(parameter);
        if (mustHighlight && hintIndex != null && hintFlags.containsKey(hintIndex)) {
            hintFlags.get(hintIndex).add(ParameterInfoUIContextEx.Flag.HIGHLIGHT);
        }
    }
}

File: src/main/kotlin/dev/xeonkryptos/xeonrobotframeworkplugin/parameterinfo/ParameterInfoHandlerUtil.kt
package dev.xeonkryptos.xeonrobotframeworkplugin.parameterinfo

import com.intellij.lang.parameterInfo.ParameterInfoUIContext
import com.intellij.lang.parameterInfo.ParameterInfoUIContextEx
import com.intellij.openapi.util.Pair
import com.intellij.psi.PsiElement
import com.intellij.psi.SyntaxTraverser
import com.jetbrains.python.documentation.PythonDocumentationProvider
import com.jetbrains.python.psi.PyCallable
import com.jetbrains.python.psi.PyFunction
import com.jetbrains.python.psi.PyNamedParameter
import com.jetbrains.python.psi.PySingleStarParameter
import com.jetbrains.python.psi.PySlashParameter
import com.jetbrains.python.psi.PyTupleParameter
import com.jetbrains.python.psi.PyUtil
import com.jetbrains.python.psi.impl.ParamHelper
import com.jetbrains.python.psi.types.PyCallableParameter
import com.jetbrains.python.psi.types.PyCallableParameterImpl
import com.jetbrains.python.psi.types.PyCallableType
import com.jetbrains.python.psi.types.PyStructuralType
import com.jetbrains.python.psi.types.PyType
import com.jetbrains.python.psi.types.TypeEvalContext
import dev.xeonkryptos.xeonrobotframeworkplugin.config.RobotOptionsProvider
import dev.xeonkryptos.xeonrobotframeworkplugin.psi.RobotTypes
import dev.xeonkryptos.xeonrobotframeworkplugin.psi.element.DefinedParameter
import dev.xeonkryptos.xeonrobotframeworkplugin.psi.element.RobotArgument
import dev.xeonkryptos.xeonrobotframeworkplugin.psi.element.RobotCallArgumentsContainer
import dev.xeonkryptos.xeonrobotframeworkplugin.psi.element.RobotParameter
import dev.xeonkryptos.xeonrobotframeworkplugin.psi.element.RobotUserKeywordStatement
import java.util.*
import java.util.function.Consumer
import java.util.function.Function
import java.util.stream.Collectors

@Suppress("UnstableApiUsage")
object ParameterInfoHandlerUtil {
    @JvmStatic
    fun <V> getCurrentParameterIndex(traverser: SyntaxTraverser<V>, offset: Int): Int {
        val root: V = traverser.getRoot()
        var curOffset = traverser.api.rangeOf(root!!).startOffset
        if (offset < curOffset) {
            return -1
        }

        var index = 0
        val delimiterTypeSet = setOf(RobotTypes.PARAMETER, RobotTypes.POSITIONAL_ARGUMENT, RobotTypes.TEMPLATE_PARAMETER, RobotTypes.TEMPLATE_ARGUMENT)
        for (child in traverser) {
            curOffset += traverser.api.rangeOf(child!!).length
            if (offset < curOffset) {
                val type = traverser.api.typeOf(child)
                if (delimiterTypeSet.contains(type)) {
                    index++
                }
                break
            }

            val type = traverser.api.typeOf(child)
            if (delimiterTypeSet.contains(type)) {
                index++
            }
        }

        return index
    }

    @JvmStatic
    fun buildHintsAndAnnotationsWithHighlights(callingFunction: PsiElement,
                                               context: ParameterInfoUIContext,
                                               currentParamIndex: Int,
                                               hintFlags: MutableMap<Int, EnumSet<ParameterInfoUIContextEx.Flag>>): Pair<MutableList<String>, MutableList<String>>? {
        val allCallArgumentsContainer = context.parameterOwner as RobotCallArgumentsContainer
        val arguments = allCallArgumentsContainer.getAllCallArguments().toTypedArray()

        val project = callingFunction.project
        val hintsAndAnnotations: Pair<MutableList<String>, MutableList<String>>
        if (callingFunction is PyFunction) {
            val indexToNamedParameter: MutableMap<Int, PyCallableParameter> = mutableMapOf()

            // param -> hint index. indexes are not contiguous, because some hints are parentheses.
            val parameterToHintIndex: MutableMap<PyCallableParameter, Int> = mutableMapOf()

            val typeEvalContext = TypeEvalContext.codeAnalysis(project, callingFunction.containingFile)
            val parameters = callingFunction.getParameters(typeEvalContext)
            val callableType = typeEvalContext.getType(callingFunction) as? PyCallableType? ?: return null

            hintsAndAnnotations = buildParameterListHint(parameters, indexToNamedParameter, parameterToHintIndex, hintFlags, typeEvalContext)
            highlightParameters(mutableListOf(*arguments), callableType, parameters, indexToNamedParameter, parameterToHintIndex, hintFlags, currentParamIndex)
        } else {
            val parameterToHintIndex: MutableMap<DefinedParameter, Int> = mutableMapOf()

            val userKeyword = callingFunction as RobotUserKeywordStatement
            val parameters = userKeyword.getInputParameters()
            val hints: MutableList<String> = ArrayList()
            val annotations: MutableList<String> = ArrayList()
            val size = parameters.size
            for ((index, parameter) in parameters.withIndex()) {
                var hint = parameter.getLookup() ?: continue
                if (parameter.hasDefaultValue()) {
                    hint += " = " + parameter.getDefaultValue()
                }
                if (index + 1 < size) {
                    hint += ", "
                }
                val hintIndex = hints.size
                hintFlags[hintIndex] = EnumSet.noneOf(ParameterInfoUIContextEx.Flag::class.java)
                parameterToHintIndex[parameter] = hintIndex
                hints.add(hint)
                annotations.add("")
            }
            hintsAndAnnotations = Pair.create(hints, annotations)
            highlightParameters(parameters.toList(), parameterToHintIndex, hintFlags, mutableListOf(*arguments), currentParamIndex)
        }

        return hintsAndAnnotations
    }

    /**
     * Builds a list of textual representation of parameters
     * Returns two lists: parameters with type hints and parameters with type annotations
     *
     * @param parameters            parameters of a callable
     * @param indexToNamedParameter used to collect all named parameters of callable
     * @param parameterToHintIndex  used to collect info about parameter hints
     * @param hintFlags             mark parameter as deprecated/highlighted/strikeout
     * @param context               context to be used to get parameter representation
     */
    fun buildParameterListHint(parameters: MutableList<PyCallableParameter>,
                               indexToNamedParameter: MutableMap<Int, PyCallableParameter>,
                               parameterToHintIndex: MutableMap<PyCallableParameter, Int>,
                               hintFlags: MutableMap<Int, EnumSet<ParameterInfoUIContextEx.Flag>>,
                               context: TypeEvalContext): Pair<MutableList<String>, MutableList<String>> {
        val hintsList: MutableList<String> = ArrayList()
        val annotations: MutableList<String> = ArrayList()
        val currentParameterIndex = intArrayOf(0)
        ParamHelper.walkDownParameters(parameters, object : ParamHelper.ParamWalker {
            override fun enterTupleParameter(param: PyTupleParameter?, first: Boolean, last: Boolean) {
                hintFlags[hintsList.size] = EnumSet.noneOf(ParameterInfoUIContextEx.Flag::class.java)
                hintsList.add("(")
                annotations.add("")
            }

            override fun leaveTupleParameter(param: PyTupleParameter?, first: Boolean, last: Boolean) {
                hintFlags[hintsList.size] = EnumSet.noneOf(ParameterInfoUIContextEx.Flag::class.java)
                hintsList.add(if (last) ")" else "), ")
                annotations.add("")
            }

            override fun visitNamedParameter(param: PyNamedParameter, first: Boolean, last: Boolean) {
                visitNonPsiParameter(PyCallableParameterImpl.psi(param), first, last)
            }

            override fun visitSlashParameter(param: PySlashParameter, first: Boolean, last: Boolean) {
                hintFlags[hintsList.size] = EnumSet.noneOf(ParameterInfoUIContextEx.Flag::class.java)
                hintsList.add(if (last) PySlashParameter.TEXT else (PySlashParameter.TEXT + ", "))
                annotations.add("")
                currentParameterIndex[0]++
            }

            override fun visitSingleStarParameter(param: PySingleStarParameter?, first: Boolean, last: Boolean) {
                hintFlags[hintsList.size] = EnumSet.noneOf(ParameterInfoUIContextEx.Flag::class.java)
                hintsList.add(if (last) PySingleStarParameter.TEXT else (PySingleStarParameter.TEXT + ", "))
                annotations.add("")
                currentParameterIndex[0]++
            }

            override fun visitNonPsiParameter(parameter: PyCallableParameter, first: Boolean, last: Boolean) {
                indexToNamedParameter[currentParameterIndex[0]] = parameter
                val stringBuilder = StringBuilder()
                var annotationAdded = false
                if (parameter.parameter is PyNamedParameter) {
                    val annotation = (parameter.parameter as PyNamedParameter).annotationValue
                    if (annotation != null) {
                        val annotationText = ParamHelper.getNameInSignature(parameter) + ": " + annotation.replace("\n", "").replace("\\s+".toRegex(), " ")
                        annotations.add(if (last) annotationText else "$annotationText, ")
                        annotationAdded = true
                    }
                }
                if (!annotationAdded) {
                    annotations.add("")
                }
                val presentableText = getPresentableText(parameter, context) { type: PyType? -> type == null || type is PyStructuralType }
                stringBuilder.append(presentableText)
                if (!last) {
                    stringBuilder.append(", ")
                }
                val hintIndex = hintsList.size
                parameterToHintIndex[parameter] = hintIndex
                hintFlags[hintIndex] = EnumSet.noneOf(ParameterInfoUIContextEx.Flag::class.java)
                hintsList.add(stringBuilder.toString())
                currentParameterIndex[0]++
            }

            // Copied from PyCallableParameter#getPresentableText(Boolean, TypeEvalContext, Predicate) as with 261.2 they changed the implementation to Kotlin which makes it byte-incompatible due to
            // the lambda expression.
            private fun getPresentableText(callableParameter: PyCallableParameter, context: TypeEvalContext, typeFilter: (PyType?) -> Boolean): String {
                if (callableParameter.parameter is PyNamedParameter || callableParameter.parameter == null) {
                    return buildString {
                        append(ParamHelper.getNameInSignature(callableParameter))

                        var renderedAsTyped = false
                        val argumentType: PyType? = callableParameter.getArgumentType(context)
                        if (!typeFilter(argumentType)) {
                            append(": ")
                            append(PythonDocumentationProvider.getTypeName(argumentType, context))
                            renderedAsTyped = true
                        }
                        append(ParamHelper.getDefaultValuePartInSignature(callableParameter.defaultValueText, renderedAsTyped) ?: "")
                    }
                }

                return PyUtil.getReadableRepr(callableParameter.parameter, false)
            }
        })
        return Pair<MutableList<String>, MutableList<String>>(hintsList, annotations)
    }

    fun highlightParameters(arguments: MutableList<RobotArgument>,
                            callableType: PyCallableType,
                            parameters: MutableList<PyCallableParameter>,
                            indexToNamedParameter: MutableMap<Int, PyCallableParameter>,
                            parameterToHintIndex: MutableMap<PyCallableParameter, Int>,
                            hintFlags: MutableMap<Int, EnumSet<ParameterInfoUIContextEx.Flag>>,
                            currentParamIndex: Int) { // gray out enough first parameters as implicit (self, cls, ...)
        for (i in 0..<callableType.implicitOffset) {
            if (indexToNamedParameter.containsKey(i)) {
                val parameter = indexToNamedParameter[i]
                hintFlags[parameterToHintIndex[parameter]]!!.add(ParameterInfoUIContextEx.Flag.DISABLE) // show but mark as absent
            }
        }

        collectHighlights(callableType, parameters, parameterToHintIndex, hintFlags, arguments, currentParamIndex)
    }

    /**
     * match params to available args, highlight current param(s)
     */
    fun collectHighlights(callableType: PyCallableType,
                          parameterList: MutableList<PyCallableParameter>,
                          parameterHintToIndex: MutableMap<PyCallableParameter, Int>,
                          hintFlags: MutableMap<Int, EnumSet<ParameterInfoUIContextEx.Flag>>,
                          args: MutableList<RobotArgument>,
                          currentParamIndex: Int) {
        val nameToCallableParameterIndex = parameterList.stream()
            .filter { param: PyCallableParameter? -> param!!.name != null }
            .collect(Collectors.toMap(Function { obj: PyCallableParameter? -> obj!!.name }, Function.identity<PyCallableParameter?>()))

        val implicitOffset = callableType.implicitOffset
        val positionalContainerIndex = findPositionalContainerIndex(parameterList)
        var parameterFound = false
        val callable: PyCallable = checkNotNull(callableType.callable)
        val parameterNameCollator = RobotOptionsProvider.getInstance(callable.project).getParameterNameCollator()

        for (i in args.indices) {
            val arg = args[i]
            val mustHighlight = currentParamIndex == i
            if (arg is RobotParameter) {
                parameterFound = true

                val parameterName = arg.getParameterName()
                if (nameToCallableParameterIndex.containsKey(parameterName)) {
                    val pyCallableParameter: PyCallableParameter = nameToCallableParameterIndex[parameterName]!!
                    highlightParameter(pyCallableParameter, parameterHintToIndex, hintFlags, mustHighlight)
                } else {
                    nameToCallableParameterIndex.keys.stream()
                        .filter { name: String? -> parameterNameCollator.equals(name, parameterName) }
                        .findFirst()
                        .map<PyCallableParameter?>(Function { key: String? -> nameToCallableParameterIndex[key] })
                        .ifPresentOrElse(Consumer { callableParameter: PyCallableParameter? ->
                            highlightParameter(callableParameter!!, parameterHintToIndex, hintFlags, mustHighlight)
                        }) {
                            val lastParameterIndex = parameterList.size - 1
                            val lastCallableParameter = parameterList[lastParameterIndex]
                            if (lastCallableParameter.isKeywordContainer) {
                                highlightParameter(lastCallableParameter, parameterHintToIndex, hintFlags, mustHighlight)
                            }
                        }
                }
            } else if (!parameterFound) {
                val argIndex = Math.clamp((i + implicitOffset).toLong(), 0, positionalContainerIndex)
                val pyCallableParameter = parameterList[argIndex]
                highlightParameter(pyCallableParameter, parameterHintToIndex, hintFlags, mustHighlight)
            }
        }
    }

    /**
     * match params to available args, highlight current param(s)
     */
    fun highlightParameters(parameterList: List<DefinedParameter>,
                            parameterHintToIndex: MutableMap<DefinedParameter, Int>,
                            hintFlags: MutableMap<Int, EnumSet<ParameterInfoUIContextEx.Flag>>,
                            args: MutableList<RobotArgument>,
                            currentParamIndex: Int) {
        val nameToCallableParameterIndex = parameterList.stream()
            .filter { param: DefinedParameter? -> param!!.getLookup() != null }
            .collect(Collectors.toMap(Function { obj: DefinedParameter? -> obj!!.getLookup() }, Function.identity<DefinedParameter?>()))

        var parameterFound = false
        for (i in args.indices) {
            val arg = args[i]
            val mustHighlight = currentParamIndex == i
            if (arg is RobotParameter) {
                parameterFound = true

                val parameterName = arg.getParameterName()
                if (nameToCallableParameterIndex.containsKey(parameterName)) {
                    val pyCallableParameter: DefinedParameter = nameToCallableParameterIndex[parameterName]!!
                    highlightParameter(pyCallableParameter, parameterHintToIndex, hintFlags, mustHighlight)
                }
            } else {
                if (!parameterFound) {
                    val pyCallableParameter = parameterList[i]
                    highlightParameter(pyCallableParameter, parameterHintToIndex, hintFlags, mustHighlight)
                }
            }
        }
    }

    private fun findPositionalContainerIndex(parameterList: MutableList<PyCallableParameter>): Int {
        for (i in parameterList.indices) {
            val parameter = parameterList[i]
            if (parameter.isPositionalContainer) {
                return i
            }
        }
        return -1
    }

    private fun highlightParameter(parameter: PyCallableParameter,
                                   parameterToHintIndex: MutableMap<PyCallableParameter, Int>,
                                   hintFlags: MutableMap<Int, EnumSet<ParameterInfoUIContextEx.Flag>>,
                                   mustHighlight: Boolean) {
        val hintIndex = parameterToHintIndex[parameter]
        if (mustHighlight && hintIndex != null && hintFlags.containsKey(hintIndex)) {
            hintFlags[hintIndex]!!.add(ParameterInfoUIContextEx.Flag.HIGHLIGHT)
        }
    }

    private fun highlightParameter(parameter: DefinedParameter,
                                   parameterToHintIndex: MutableMap<DefinedParameter, Int>,
                                   hintFlags: MutableMap<Int, EnumSet<ParameterInfoUIContextEx.Flag>>,
                                   mustHighlight: Boolean) {
        val hintIndex = parameterToHintIndex[parameter]
        if (mustHighlight && hintIndex != null && hintFlags.containsKey(hintIndex)) {
            hintFlags[hintIndex]!!.add(ParameterInfoUIContextEx.Flag.HIGHLIGHT)
        }
    }
}
